include/generic_datatype.h

```diff
@@ -1,16 +1,77 @@
 
-#pragma once
+//#pragma once
 
-#include <functional>
+//#include <functional>
+
+#include <unordered_map>
+#include <vector>
 
 #include "datatype.h"
 #include "exceptions.h"
+#include "generic_sort.h"
 #include "smt_defs.h"
-#include "sort.h"
 
 // using namespace smt;
 namespace smt {
 
+struct selectorComponents
+{
+  std::string name;
+  Sort sort;
+};
+
+class DatatypeMemoryManager
```

Comment:
I think we should try to be similar to `generic_sort.h` and `generic_term.h`. The main class of this file should be `GenericDatatype` and should inherit from `AbsDatatype`. Such a class represents a single datatype, no a global store of datatypes. More comments about this below.